core-bundle/src/Filesystem/Filesystem.php

```diff
@@ -0,0 +1,416 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Filesystem;
+
+use Contao\StringUtil;
+use Contao\Validator;
+use League\Flysystem\Config;
+use League\Flysystem\DirectoryAttributes;
+use League\Flysystem\DirectoryListing;
+use League\Flysystem\FileAttributes;
+use League\Flysystem\FilesystemAdapter;
+use League\Flysystem\FilesystemOperator;
+use League\Flysystem\InvalidStreamProvided;
+use League\Flysystem\PathNormalizer;
+use League\Flysystem\StorageAttributes;
+use League\Flysystem\WhitespacePathNormalizer;
+use ProxyManager\Factory\LazyLoadingGhostFactory;
+use ProxyManager\Proxy\GhostObjectInterface;
+
+/**
+ * This Flysystem operator wraps a DBAFS layer around the underlying filesystem
+ * adaptor: Listing content will only access the database, operations that can
+ * mutate the filesystem, will automatically trigger a synchronization.
+ *
+ * Additionally, there is support to directly retrieve the 'extra metadata'
+ * from a file's / directory's StorageAttributes via a dedicated method.
+ *
+ * We widen the FilesystemOperator interface - instead of calling methods with
+ * resource paths (string), you can also pass in an ID (int) or UUID of a
+ * synchronized DBAFS resource.
+ *
+ * @phpstan-import-type ExtraMetadata from Dbafs
+ */
+final class Filesystem implements FilesystemOperator
```

Comment:
The biggest problem I have with this is the name. Our extension developers will associate `Filesystem` with the local drive for sure. Especially because it's already taken by Symfony's `Filesystem` class which indeed provides methods to operate on the local drive. What about we call it what it really is `VirtualFilesystem`? It would also make a lot of things clearer.

Reply:
Thought about this as well and just went with `Filesystem` because Flysystem did. 

I like `VirtualFilesystem`.

Alternatives would maybe be:
* `Flysystem`
* `FilesystemOperator`
* `FlysystemOperator`

We might also want to reword `Dbafs` in this go because it's less of a **f**ile**s**ystem but rather a "metadata storage indexed by paths".

Reply:
I don't care about `Dbafs` that much because it's an internal implementation detail now :) Which is great by the way!
I prefer `VirtualFilesystem` over the others. I'm trying to imagine the users of that service and for them, I think `VirtualFilesystem` tells a lot more than `Flysystem` (does it fly?) or `FlysystemOperator` (so it not only flies but it also has an operator that flies it?).
Of course we'll have devs that know exactly what `Flysystem` is. But they'll also understand `VirtualFilesystem` ;)

Reply:
I started outlining a mount manager concept. I will update the description once there is more to see and then we can discuss based on this new approach.